File: app/src/main/java/Model/PropertyTemplate.java
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017-2020 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package Model;

import android.content.SharedPreferences;

import org.ksoap2.serialization.KvmSerializable;
import org.ksoap2.serialization.PropertyInfo;
import org.ksoap2.serialization.SoapObject;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Hashtable;
import java.util.Locale;

import androidx.annotation.NonNull;

public class PropertyTemplate  extends SoapableObject implements Comparable<PropertyTemplate>, Serializable, KvmSerializable {
    private static final long serialVersionUID = 1L;

    private final static int ID_NEW = -1;
    private final static int ID_MRU = -2;
    private final static int ID_SIM = -3;
    private final static int ID_ANON = -4;

    final static int GROUP_AUTO = 0;

    public static PropertyTemplate[] sharedTemplates;

    public static final String PREF_KEY_TEMPLATES = "preferenceKeyTemplates";

    private enum templatePropID {pidTemplateID, pidTemplateName, pidTemplateDescription, pidTemplateGroupAsInt, pidTemplateGroupDisplayName, pidTemplatePropertyTypes, pidTemplateDefault}

    // region basic
    private int ID;
    public String Name;
    public String Description;
    int GroupAsInt;
    String GroupDisplayName;
    private HashSet<Integer> PropertyTypes;
    private Boolean IsDefault;

    private void init() {
        ID = -1;
        Name = Description = GroupDisplayName = "";
        PropertyTypes = new HashSet<>();
    }

    public PropertyTemplate() {
        init();
    }

    public PropertyTemplate(SoapObject so) {
        super();
        init();
        FromProperties(so);
    }

    @NonNull
    @Override
    public String toString() {
        return String.format(Locale.getDefault(), "%d: (%s) %s %s", ID, GroupDisplayName, Name, IsDefault ? " (default)" : "");
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof PropertyTemplate))
            return false;
        if (o == this)
            return true;
        return ID == ((PropertyTemplate)o).ID;
    }

    @Override
    public int hashCode() {
        return ((Integer)ID).hashCode();
    }

    @Override
    public int compareTo(@NonNull PropertyTemplate o) {
        if (this.GroupAsInt == o.GroupAsInt) {
            if (this.GroupAsInt == GROUP_AUTO)    // Automatic groups are always first - so if we're here, then both of us are automatic; resolve on ID inverse
                return Integer.compare(o.ID, this.ID);
            else
                return this.Name.compareTo(o.Name);
        }
        return Integer.compare(this.GroupAsInt, o.GroupAsInt);
    }
    // endregion

    // region SOAP
    @Override
    public void ToProperties(SoapObject so) {
        so.addProperty("ID", ID);
        so.addProperty("Name", Name);
        so.addProperty("Description", Description);
        so.addProperty("GroupAsInt", GroupAsInt);
        so.addProperty("GroupDisplayName", GroupDisplayName);
        so.addProperty("PropertyTypes", PropertyTypes.toArray());
        so.addProperty("IsDefault", IsDefault);
    }

    @Override
    protected void FromProperties(SoapObject so) {
        ID = Integer.parseInt(so.getProperty("ID").toString());
        Name = so.getProperty("Name").toString();
        Description = so.getProperty("Description").toString();
        GroupAsInt = Integer.parseInt(so.getProperty("GroupAsInt").toString());
        GroupDisplayName = so.getProperty("GroupDisplayName").toString();
        PropertyTypes.clear();
        SoapObject types = (SoapObject) so.getProperty("PropertyTypes");
        int cVals = types.getPropertyCount();
        for (int i = 0; i < cVals; i++)
            PropertyTypes.add(Integer.parseInt(types.getPropertyAsString(i)));
        IsDefault = Boolean.parseBoolean(so.getProperty("IsDefault").toString());
    }

    public Object getProperty(int arg0) {
        templatePropID pid = templatePropID.values()[arg0];
        switch (pid) {
            case pidTemplateID:
                return this.ID;
            case pidTemplateName:
                return this.Name;
            case pidTemplateDescription:
                return this.Description;
            case pidTemplateGroupAsInt:
                return this.GroupAsInt;
            case pidTemplateGroupDisplayName:
                return this.GroupDisplayName;
            case pidTemplatePropertyTypes:
                return this.PropertyTypes;
            case pidTemplateDefault:
                return this.IsDefault;
        }
        return null;
    }

    public int getPropertyCount() {
        return templatePropID.values().length;
    }

    public void getPropertyInfo(int arg0, Hashtable arg1, PropertyInfo pi) {
        templatePropID pid = templatePropID.values()[arg0];
        switch (pid) {
            case pidTemplateID:
                pi.type = PropertyInfo.INTEGER_CLASS;
                pi.name = "ID";
            case pidTemplateName:
                pi.type = PropertyInfo.STRING_CLASS;
                pi.name = "Name";
            case pidTemplateDescription:
                pi.type = PropertyInfo.STRING_CLASS;
                pi.name = "Description";
            case pidTemplateGroupAsInt:
                pi.type = PropertyInfo.INTEGER_CLASS;
                pi.name = "GroupAsInt";
            case pidTemplateGroupDisplayName:
                pi.type = PropertyInfo.STRING_CLASS;
                pi.name = "GroupDisplayName";
            case pidTemplatePropertyTypes:
                pi.type = PropertyInfo.VECTOR_CLASS;
                pi.name = "PropertyTypes";
                pi.elementType = new PropertyInfo();
                pi.elementType.type = PropertyInfo.INTEGER_CLASS;
                pi.elementType.name = "integer";
            case pidTemplateDefault:
                pi.type = PropertyInfo.BOOLEAN_CLASS;
                pi.name = "IsDefault";
            default:
                break;
        }
    }

    public void setProperty(int arg0, Object arg1) {
        templatePropID pid = templatePropID.values()[arg0];
        String sz = arg1.toString();
        switch (pid) {
            case pidTemplateID:
                ID = Integer.parseInt(sz);
            case pidTemplateName:
                Name = sz;
            case pidTemplateDescription:
                Description = sz;
            case pidTemplateGroupAsInt:
                GroupAsInt = Integer.parseInt(sz);
            case pidTemplateGroupDisplayName:
                GroupDisplayName = sz;
            case pidTemplatePropertyTypes:
                PropertyTypes.clear();
                int[] rgVals = (int[]) arg1;
                for (int i : rgVals)
                    PropertyTypes.add(i);
            case pidTemplateDefault:
                IsDefault = Boolean.parseBoolean(sz);
            default:
                break;
        }
    }
    // endregion

    // region Retrieving/persistance
    private static final String prefSharedTemplates = "prefsSharedTemplates";

    public static void saveSharedTemplates(SharedPreferences pref) {
        SharedPreferences.Editor e = pref.edit();
        e.putString(prefSharedTemplates, MFBUtil.serializeToString(sharedTemplates));
        e.apply();
        e.commit();
    }

    public static PropertyTemplate[] getSharedTemplates(SharedPreferences pref) {
        String s = pref.getString(prefSharedTemplates, "");
        sharedTemplates = MFBUtil.deserializeFromString(s);
        return sharedTemplates;
    }

    public static PropertyTemplate[] getDefaultTemplates() {
        ArrayList<PropertyTemplate> al = new ArrayList<>();
        for (PropertyTemplate pt : PropertyTemplate.sharedTemplates) {
            if (pt.IsDefault)
                al.add(pt);
        }
        return al.toArray(new PropertyTemplate[0]);
    }

    private static PropertyTemplate templateWithID(int id) {
        if (sharedTemplates != null) {
            for (PropertyTemplate pt : sharedTemplates) {
                if (pt.ID == id)
                    return pt;
            }
        }
        return null;
    }

    public static PropertyTemplate[] templatesWithIDs(HashSet<Integer> rgid) {
        ArrayList<PropertyTemplate> al = new ArrayList<>();
        if (rgid != null) {
            for (int i : rgid)
                al.add(PropertyTemplate.templateWithID(i));
        }
        return al.toArray(new PropertyTemplate[0]);
    }

    public static PropertyTemplate getAnonTemplate() { return templateWithID(ID_ANON); }

    public static PropertyTemplate getSimTemplate() { return templateWithID(ID_SIM); }
    // endregion

    // region merging
    public static HashSet<Integer> mergeTemplates(PropertyTemplate[] rgpt) {
        HashSet<Integer> result = new HashSet<>();
        if (rgpt != null) {
            for (PropertyTemplate pt : rgpt)
                if (pt != null && pt.PropertyTypes != null)
                    result.addAll(pt.PropertyTypes);
        }
        return result;
    }
    // endregion

}

File: app/src/main/java/model/PropertyTemplate.kt
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017-2025 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package model

import android.content.SharedPreferences
import model.MFBUtil.deserializeFromString
import model.MFBUtil.serializeToString
import org.ksoap2.serialization.KvmSerializable
import org.ksoap2.serialization.PropertyInfo
import org.ksoap2.serialization.SoapObject
import java.io.Serializable
import java.util.*

class PropertyTemplate(so: SoapObject) : SoapableObject(), Comparable<PropertyTemplate>, Serializable,
    KvmSerializable {
    private enum class TemplatePropID {
        PIDTemplateID, PIDTemplateName, PIDTemplateDescription, PIDTemplateGroupAsInt, PIDTemplateGroupDisplayName, PIDTemplatePropertyTypes, PIDTemplateDefault
    }

    // region basic
    private var id = 0
    @JvmField
    var name: String = ""
    @JvmField
    var description: String = ""
    @JvmField
    var groupAsInt = 0
    @JvmField
    var groupDisplayName: String = ""
    private var propTypes: HashSet<Int>? = null
    private var isDefault: Boolean? = null
    private fun init() {
        id = -1
        groupDisplayName = ""
        description = groupDisplayName
        name = description
        propTypes = HashSet()
    }

    init {
        init()
        fromProperties(so)
    }

    override fun toString(): String {
        return String.format(
            Locale.getDefault(),
            "%d: (%s) %s %s",
            id,
            groupDisplayName,
            name,
            if (isDefault!!) " (default)" else ""
        )
    }

    override fun equals(other: Any?): Boolean {
        if (other !is PropertyTemplate) return false
        return if (other === this) true else id == other.id
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    override fun compareTo(other: PropertyTemplate): Int {
        return if (groupAsInt == other.groupAsInt) {
            if (groupAsInt == GROUP_AUTO) // Automatic groups are always first - so if we're here, then both of us are automatic; resolve on ID inverse
                other.id.compareTo(id) else name.compareTo(other.name)
        } else groupAsInt.compareTo(other.groupAsInt)
    }

    // endregion
    // region SOAP
    override fun toProperties(so: SoapObject) {
        so.addProperty("ID", id)
        so.addProperty("Name", name)
        so.addProperty("Description", description)
        so.addProperty("GroupAsInt", groupAsInt)
        so.addProperty("GroupDisplayName", groupDisplayName)
        so.addProperty("PropertyTypes", propTypes!!.toTypedArray())
        so.addProperty("IsDefault", isDefault)
    }

    override fun fromProperties(so: SoapObject) {
        id = so.getProperty("ID").toString().toInt()
        name = so.getProperty("Name").toString()
        description = so.getPrimitivePropertyAsString("Description").toString()
        groupAsInt = so.getProperty("GroupAsInt").toString().toInt()
        groupDisplayName = so.getProperty("GroupDisplayName").toString()
        propTypes!!.clear()
        val types = so.getProperty("PropertyTypes") as SoapObject
        val cVals = types.propertyCount
        for (i in 0 until cVals) propTypes!!.add(types.getPropertyAsString(i).toInt())
        isDefault = java.lang.Boolean.parseBoolean(so.getProperty("IsDefault").toString())
    }

    override fun getProperty(arg0: Int): Any {
        return when (TemplatePropID.values()[arg0]) {
            TemplatePropID.PIDTemplateID -> id
            TemplatePropID.PIDTemplateName -> name
            TemplatePropID.PIDTemplateDescription -> description
            TemplatePropID.PIDTemplateGroupAsInt -> groupAsInt
            TemplatePropID.PIDTemplateGroupDisplayName -> groupDisplayName
            TemplatePropID.PIDTemplatePropertyTypes -> propTypes!!
            TemplatePropID.PIDTemplateDefault -> isDefault!!
        }
    }

    override fun getPropertyCount(): Int {
        return TemplatePropID.values().size
    }

    override fun getPropertyInfo(arg0: Int, arg1: Hashtable<*, *>?, pi: PropertyInfo) {
        when (TemplatePropID.values()[arg0]) {
            TemplatePropID.PIDTemplateID -> {
                pi.type = PropertyInfo.INTEGER_CLASS
                pi.name = "ID"
                }
            TemplatePropID.PIDTemplateName -> {
                pi.type = PropertyInfo.STRING_CLASS
                pi.name = "Name"
            }
            TemplatePropID.PIDTemplateDescription -> {
                pi.type = PropertyInfo.STRING_CLASS
                pi.name = "Description"
            }
            TemplatePropID.PIDTemplateGroupAsInt -> {
                pi.type = PropertyInfo.INTEGER_CLASS
                pi.name = "GroupAsInt"
            }
            TemplatePropID.PIDTemplateGroupDisplayName -> {
                pi.type = PropertyInfo.STRING_CLASS
                pi.name = "GroupDisplayName"
            }
            TemplatePropID.PIDTemplatePropertyTypes -> {
                pi.type = PropertyInfo.VECTOR_CLASS
                pi.name = "PropertyTypes"
                pi.elementType = PropertyInfo()
                pi.elementType.type = PropertyInfo.INTEGER_CLASS
                pi.elementType.name = "integer"
            }
            TemplatePropID.PIDTemplateDefault -> {
                pi.type = PropertyInfo.BOOLEAN_CLASS
                pi.name = "IsDefault"
            }
        }
    }

    override fun setProperty(arg0: Int, arg1: Any) {
        val pid = TemplatePropID.values()[arg0]
        val sz = arg1.toString()
        when (pid) {
            TemplatePropID.PIDTemplateID -> {
                id = sz.toInt()
                name = sz
                description = sz
                groupAsInt = sz.toInt()
                groupDisplayName = sz
                propTypes!!.clear()
                val rgVals = arg1 as IntArray
                for (i in rgVals) propTypes!!.add(i)
                isDefault = java.lang.Boolean.parseBoolean(sz)
            }
            TemplatePropID.PIDTemplateName -> {
                name = sz
                description = sz
                groupAsInt = sz.toInt()
                groupDisplayName = sz
                propTypes!!.clear()
                val rgVals = arg1 as IntArray
                for (i in rgVals) propTypes!!.add(i)
                isDefault = java.lang.Boolean.parseBoolean(sz)
            }
            TemplatePropID.PIDTemplateDescription -> {
                description = sz
                groupAsInt = sz.toInt()
                groupDisplayName = sz
                propTypes!!.clear()
                val rgVals = arg1 as IntArray
                for (i in rgVals) propTypes!!.add(i)
                isDefault = java.lang.Boolean.parseBoolean(sz)
            }
            TemplatePropID.PIDTemplateGroupAsInt -> {
                groupAsInt = sz.toInt()
                groupDisplayName = sz
                propTypes!!.clear()
                val rgVals = arg1 as IntArray
                for (i in rgVals) propTypes!!.add(i)
                isDefault = java.lang.Boolean.parseBoolean(sz)
            }
            TemplatePropID.PIDTemplateGroupDisplayName -> {
                groupDisplayName = sz
                propTypes!!.clear()
                val rgVals = arg1 as IntArray
                for (i in rgVals) propTypes!!.add(i)
                isDefault = java.lang.Boolean.parseBoolean(sz)
            }
            TemplatePropID.PIDTemplatePropertyTypes -> {
                propTypes!!.clear()
                val rgVals = arg1 as IntArray
                for (i in rgVals) propTypes!!.add(i)
                isDefault = java.lang.Boolean.parseBoolean(sz)
            }
            TemplatePropID.PIDTemplateDefault -> isDefault = java.lang.Boolean.parseBoolean(sz)
        }
    }

    companion object {
        @Suppress("UNUSED")
        private const val serialVersionUID = 1L
        @Suppress("UNUSED")
        private const val ID_NEW = -1
        @Suppress("UNUSED")
        private const val ID_MRU = -2
        private const val ID_SIM = -3
        private const val ID_ANON = -4
        const val GROUP_AUTO = 0
        @JvmField
        var sharedTemplates: Array<PropertyTemplate>? = null
        const val PREF_KEY_TEMPLATES = "preferenceKeyTemplates"

        // endregion
        // region Retrieving/persistance
        private const val prefSharedTemplates = "prefsSharedTemplates"
        fun saveSharedTemplates(pref: SharedPreferences) {
            val e = pref.edit()
            e.putString(prefSharedTemplates, serializeToString(sharedTemplates))
            e.apply()
        }

        @JvmStatic
        fun getSharedTemplates(pref: SharedPreferences): Array<PropertyTemplate>? {
            val s = pref.getString(prefSharedTemplates, "")
            sharedTemplates = deserializeFromString<Array<PropertyTemplate>>(s)
            return sharedTemplates
        }

        @JvmStatic
        val defaultTemplates: Array<PropertyTemplate>
            get() {
                val al = ArrayList<PropertyTemplate>()
                for (pt in sharedTemplates!!) {
                    if (pt.isDefault!!) al.add(pt)
                }
                return al.toTypedArray()
            }

        private fun templateWithID(id: Int): PropertyTemplate? {
            if (sharedTemplates != null) {
                for (pt in sharedTemplates!!) {
                    if (pt.id == id) return pt
                }
            }
            return null
        }

        @JvmStatic
        fun templatesWithIDs(rgid: HashSet<Int>?): Array<PropertyTemplate> {
            val al = ArrayList<PropertyTemplate>()
            if (rgid != null) {
                for (i in rgid) {
                    val template = templateWithID(i)
                    if (template != null)
                        al.add(template)
                }
            }
            return al.toTypedArray()
        }

        @JvmStatic
        val anonTemplate: PropertyTemplate?
            get() = templateWithID(ID_ANON)
        @JvmStatic
        val simTemplate: PropertyTemplate?
            get() = templateWithID(ID_SIM)

        // endregion
        // region merging
        @JvmStatic
        fun mergeTemplates(rgpt: Array<PropertyTemplate?>?): HashSet<Int> {
            val result = HashSet<Int>()
            if (rgpt != null) {
                for (pt in rgpt) if (pt?.propTypes != null) result.addAll(
                    pt.propTypes!!
                )
            }
            return result
        } // endregion
    }
}